File: src/content/algorithms/bfs/code.py
from collections import deque  # #@init-queue


def bfs(graph, start):  # #@function-def
    visited = {start}  # #@init-visited
    queue = deque([start])  # #@init-queue
    result = []

    while queue:  # #@main-loop
        vertex = queue.popleft()  # #@dequeue
        result.append(vertex)  # #@process-node

        for neighbor in graph[vertex]:  # #@enqueue
            if neighbor not in visited:  # #@check-visited
                visited.add(neighbor)  # #@mark-visited
                queue.append(neighbor)  # #@enqueue

    return result


def bfs_with_distance(graph, start):  # #@function-def
    visited = {start}  # #@init-visited
    queue = deque([(start, 0)])  # #@init-queue
    distances = {start: 0}  # #@track-distance

    while queue:  # #@main-loop
        vertex, dist = queue.popleft()  # #@dequeue

        for neighbor in graph[vertex]:  # #@enqueue
            if neighbor not in visited:  # #@check-visited
                visited.add(neighbor)  # #@mark-visited
                distances[neighbor] = dist + 1  # #@track-distance
                queue.append((neighbor, dist + 1))  # #@enqueue

    return distances

File: src/content/algorithms/bfs/code.rs
use std::collections::{HashMap, HashSet, VecDeque};  // #@init-queue

pub fn bfs(  // #@function-def
    graph: &HashMap<char, Vec<char>>,
    start: char,
) -> Vec<char> {
    let mut visited = HashSet::new();  // #@init-visited
    visited.insert(start);
    let mut queue = VecDeque::new();  // #@init-queue
    queue.push_back(start);
    let mut result = Vec::new();

    while let Some(vertex) = queue.pop_front() {  // #@main-loop  // #@dequeue
        result.push(vertex);  // #@process-node

        if let Some(neighbors) = graph.get(&vertex) {  // #@enqueue
            for &neighbor in neighbors {
                if !visited.contains(&neighbor) {  // #@check-visited
                    visited.insert(neighbor);  // #@mark-visited
                    queue.push_back(neighbor);  // #@enqueue
                }
            }
        }
    }

    result
}

pub fn bfs_with_distance(  // #@function-def
    graph: &HashMap<char, Vec<char>>,
    start: char,
) -> HashMap<char, u32> {
    let mut visited = HashSet::new();  // #@init-visited
    visited.insert(start);
    let mut queue = VecDeque::new();  // #@init-queue
    queue.push_back((start, 0u32));
    let mut distances = HashMap::new();  // #@track-distance
    distances.insert(start, 0);

    while let Some((vertex, dist)) = queue.pop_front() {  // #@main-loop  // #@dequeue
        if let Some(neighbors) = graph.get(&vertex) {  // #@enqueue
            for &neighbor in neighbors {
                if !visited.contains(&neighbor) {  // #@check-visited
                    visited.insert(neighbor);  // #@mark-visited
                    distances.insert(neighbor, dist + 1);  // #@track-distance
                    queue.push_back((neighbor, dist + 1));  // #@enqueue
                }
            }
        }
    }

    distances
}
